### PR TITLE
[Refactor][Manifest] drop generative-op carrier; allow empty signature.inputs

### LINKED ISSUE #1440: [REFACTOR][MANIFEST] drop generative-op carrier-scalar; allow empty signature.inputs
### Symptom / Motivation
The current manifest schema enforces `len(signature.inputs) >= 1` via `test_every_signature_has_inputs_and_outputs`. Two reduction-family ops (`AlibiFwdOp`, `SinusoidalFwdOp`) synthesize their output entirely from construction-time scalar parameters — they take zero tensor inputs. To satisfy the schema, both manifest entries declare a fake `device_carrier` 0-dim scalar tensor as their sole input; the carrier exists only as a workaround. This same carrier is also threaded through `_register_generative_custom_op` and the validator carries a `is_generative` carve-out that detects the pattern via `ref_api == "none" + 0 positional args` and skips L1/C4 alignment checks. The hack documents itself in the manifest YAML preamble; the design doc previously enshrined it as well (removed in #1439).

The real defect is one schema invariant being too strict, not a missing "generative op kind". Every op has inputs — they are not always tensors. ALiBi and sinusoidal have integer + dtype params as their real inputs.

### Root Cause Analysis
- `scripts/validate_manifest.py:test_every_signature_has_inputs_and_outputs` requires `len(signature.inputs) >= 1`. Generative ops physically violate this.
- `scripts/validate_manifest.py` threads `is_generative` through `check_l1` and L4 (lines around 542, 836-862, 3398-3408) to skip the inputs↔forward positional alignment. The flag is derived from a `ref_api == "none" + 0 positional` heuristic.
- `tileops/ops/elementwise/_base.py:_register_generative_custom_op` registers the op with `device_carrier: Tensor` as its first arg so torch.compile / `register_fake` has a tensor to trace.
- `tileops/ops/elementwise/alibi.py` and `tileops/ops/elementwise/sinusoidal.py` allocate `self._device_carrier = torch.empty((), dtype=dtype, device="cuda")` in `__init__` and pass it to the custom op call in `forward()`.
- `tileops/manifest/elementwise_generative.yaml` declares `device_carrier` as the sole input on both entries; output dtype uses `same_as(device_carrier)` to channel dtype through the carrier.

The cleanup vector is: relax the schema invariant, drop the carrier and its plumbing, simplify the validator.

### Related Files
- `scripts/validate_manifest.py`
- `tileops/ops/elementwise/_base.py`
- `tileops/ops/elementwise/alibi.py`
- `tileops/ops/elementwise/sinusoidal.py`
- `tileops/manifest/elementwise_generative.yaml`
- `tests/test_validate_manifest.py`
- `tests/ops/test_special_elementwise.py`

## Goal
Replace the carrier-scalar hack with the minimum-correct implementation that allows `signature.inputs: {}` for ops whose output is fully derived from construction-time params. After this change, no "generative-op" carve-out concept exists anywhere — the schema is permissive enough that the alignment checks become natural no-ops on empty input lists.

## Plan

1. Relax `scripts/validate_manifest.py:test_every_signature_has_inputs_and_outputs` from `len(signature.inputs) >= 1` to `len(signature.outputs) >= 1 AND (len(signature.inputs) >= 1 OR len(signature.params) >= 1)`. Add a test that an op with `inputs: {}` plus non-empty `params` is accepted.
2. Remove the `is_generative` parameter from `check_l1` and L4 in `scripts/validate_manifest.py`. Remove the `ref_api == "none" + 0 positional` heuristic detection (around lines 836-862 and 3398-3408). The inputs↔forward positional alignment loop becomes a natural no-op when `inputs == []`.
3. Remove all "Generative-op carve-out" comments from `scripts/validate_manifest.py`.
4. Update `tileops/manifest/elementwise_generative.yaml`: change `inputs: { device_carrier: ... }` to `inputs: {}` on both entries. Promote `dtype` from a fake `same_as(device_carrier)` channel to an explicit `params: dtype: {type: torch.dtype}` entry. Update output dtype expression accordingly (e.g., `same_as(dtype)` or direct reference). Delete the file's preamble comment block describing the carve-out.
5. Replace `tileops/ops/elementwise/_base.py:_register_generative_custom_op` with a tensor-input-less impl. Two acceptable approaches: (a) drop the custom_op registration entirely and have `forward()` call the kernel directly (eager-only; the ops cannot be `torch.compile`-graph-captured); or (b) register via `torch.library.impl` directly without going through dispatcher tracing. Pick (a) for this PR unless graph capture is required by a downstream consumer.
6. Delete `self._device_carrier` allocation in `tileops/ops/elementwise/alibi.py:__init__` and `tileops/ops/elementwise/sinusoidal.py:__init__`. Simplify `forward()` to call the kernel directly without passing a carrier.
7. Delete validator tests that assert the carve-out fires (in `tests/test_validate_manifest.py`); add one test confirming `inputs: {}` is accepted by the relaxed invariant.
8. Run `pytest tests/ops/test_special_elementwise.py` to confirm `AlibiFwdOp` / `SinusoidalFwdOp` still produce correct output without the carrier path.
9. Run `python scripts/validate_manifest.py` and `pytest tests/test_validate_manifest.py` — both pass.

## Constraints
- Joint manifest + validator + ops change. The schema relaxation and the impl changes that depend on it land in the same PR — splitting them would leave the manifest mid-state for one round.
- MUST NOT touch other manifest entries that currently use `ref_api: "none"` with real tensor inputs (the 11 attention.yaml entries and 3 elementwise_fused_gated.yaml entries). They are not affected.
- MUST NOT introduce a new "generative op kind" concept anywhere. The fix is schema relaxation, not categorization.
- MUST keep `AlibiFwdOp` / `SinusoidalFwdOp` working end-to-end. `tests/ops/test_special_elementwise.py` must stay green.
- Dropping torch.compile graph capture for ALiBi / sinusoidal is acceptable in this PR. If a downstream consumer requires it, file a separate follow-up; do not solve it here with another carrier-style placeholder.

## Acceptance Criteria
- [ ] Modified files pass unit tests (`pytest tests/test_validate_manifest.py tests/ops/test_special_elementwise.py`, all green on CUDA).
- [ ] `python scripts/validate_manifest.py` exits 0.
- [ ] `grep -rn 'device_carrier\|_register_generative_custom_op\|is_generative\|Generative-op carve-out\|generative-op carve-out' tileops/ scripts/ tests/` returns no matches (every trace of the hack is gone).
- [ ] `grep -rn 'signature.inputs' scripts/validate_manifest.py tests/test_validate_manifest.py` shows no `>= 1` invariant; the relaxed `outputs >= 1 AND (inputs >= 1 OR params >= 1)` is in place.
- [ ] `tileops/manifest/elementwise_generative.yaml` declares `inputs: {}` on `AlibiFwdOp` and `SinusoidalFwdOp`; `dtype` is a `param`; the file preamble has no carve-out narrative.
- [ ] A new test `test_signature_inputs_may_be_empty_when_params_present` (or equivalent) is added to `tests/test_validate_manifest.py` and passes.

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #1440

## Summary

- Relax `validate_manifest.py`: `outputs >= 1 AND (inputs >= 1 OR params >= 1)` replaces the old `inputs >= 1` invariant; the `is_generative` parameter and `ref_api == 'none'` heuristic are removed from `check_l1` and L4.
- `tileops/manifest/elementwise_generative.yaml`: `AlibiFwdOp` and `SinusoidalFwdOp` now declare `inputs: {}` with `dtype` promoted to a real `params:` entry; the carve-out preamble is gone.
- `tileops/ops/elementwise/_base.py`: drop `_register_generative_custom_op`; `alibi.py` / `sinusoidal.py` drop the `self._device_carrier` allocation and call the kernel directly (eager-only — torch.compile graph capture deferred to a follow-up if needed).
- Add `test_signature_inputs_may_be_empty_when_params_present` in `tests/test_validate_manifest.py` exercising the relaxed schema.
- No `device_carrier` / `_register_generative_custom_op` / `is_generative` / `generative-op carve-out` tokens remain anywhere in `tileops/`, `scripts/`, `tests/`.

## Test plan

- [x] pytest tests/test_validate_manifest.py tests/ops/test_special_elementwise.py — 301 passed, 3 warnings, 0 failed
- [x] python scripts/validate_manifest.py — exits 0, "All manifest checks passed."
- [x] grep `device_carrier|_register_generative_custom_op|is_generative|Generative-op carve-out|generative-op carve-out` on `tileops/ scripts/ tests/` — no matches
- [x] grep `signature.inputs` on validator + tests — no `>= 1` invariant remains; relaxed form is in place
- [x] Manifest YAML check: both ops have `inputs: {}` and `dtype` as a param; preamble carve-out narrative removed
- [x] New test `test_signature_inputs_may_be_empty_when_params_present` present and passing

## Test node delta

```
File                               Base    HEAD    Delta
--------------------------------------------------------
tests/test_validate_manifest.py     218     220       +2
--------------------------------------------------------
TOTAL                               218     220       +2

Growth: +0.9%
```

**Justification:** one new schema test (`test_signature_inputs_may_be_empty_when_params_present`) is added to lock in the relaxed `outputs >= 1 AND (inputs >= 1 OR params >= 1)` invariant required by AC-6, and one ancillary node tracks the symmetric `params`-less rejection path. AC-6 mandates this addition; without it the relaxation is unverified.

### CODE REVIEW

File: tileops/ops/elementwise/alibi.py [hunk 1]
```diff
@@ -25,7 +25,6 @@ class AlibiFwdOp(Op):
     """
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/elementwise/alibi.py:32]: Adopted. Added an eager-only / no-torch.compile-wrapping Note to the `AlibiFwdOp` docstring. See 40bf0bb.

File: tileops/ops/elementwise/sinusoidal.py [hunk 2]
```diff
@@ -25,7 +25,6 @@ class SinusoidalFwdOp(Op):
     """
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/elementwise/sinusoidal.py:32]: Adopted. Added an eager-only / no-torch.compile-wrapping Note to the `SinusoidalFwdOp` docstring. See 40bf0bb.

File: tileops/ops/elementwise/__init__.py [hunk 3]
```diff
@@ -276,19 +275,5 @@
 # correctly for both same-shape and broadcasting inputs.
 _register_lerp_tensor_custom_op(LerpTensorFwdOp)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/elementwise/__init__.py:280]: Adopted. Updated the registration-section comment to spell out that `AlibiFwdOp` / `SinusoidalFwdOp` are intentionally excluded (zero tensor inputs, eager-only). See 40bf0bb.

File: tests/test_validate_manifest.py [hunk 4]
```diff
@@ -531,6 +531,45 @@ def test_shape_rule_unknown_callable_dedup_per_rule(self, validator):
...
+        )
+        errors = validator.check_l0("vacuous_op", entry)
+        assert any(
+            "[schema]" in e and "input" in e and "param" in e
+            for e in errors
+        ), (
+            f"Expected schema error when both inputs and params are empty, "
+            f"got: {errors}"
+        )
+
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/test_validate_manifest.py:588]: Adopted. Added `test_signature_outputs_empty_fails` covering the L0 schema check that `signature.outputs` must declare at least one tensor. See 15d40e2.

File: scripts/validate_manifest.py [hunk 5]
```diff
@@ -279,6 +279,29 @@ def check_l0(
...
+        # ops synthesize the output from params alone); an op with neither
+        # inputs nor params has no surface the validator or codegen can act
+        # on.
+        raw_inputs = sig.get("inputs")
+        raw_outputs = sig.get("outputs")
+        raw_params = sig.get("params")
+        inputs_count = len(raw_inputs) if isinstance(raw_inputs, dict) else 0
+        outputs_count = len(raw_outputs) if isinstance(raw_outputs, dict) else 0
+        params_count = len(raw_params) if isinstance(raw_params, dict) else 0
+        if outputs_count < 1:
```
Comment (Ibuki-wind) [inline scripts/validate_manifest.py:294]: New-path coverage fails here: `check_l0` now rejects `outputs={}` through the new `outputs_count < 1` branch, but the test suite never exercises an empty-outputs manifest, so this invariant can regress silently → add a `validator.check_l0(..., _make_entry(outputs={}))` case that asserts the new `[schema] ... signature.outputs` error.

File: tileops/ops/elementwise/__init__.py [hunk 6]
```diff
@@ -276,19 +279,5 @@
...
 
-# --- Generative ops (2 ops: no tensor input -> out) ---
```
Comment (Ibuki-wind) [inline tileops/ops/elementwise/__init__.py:279]: The deleted generative registration block makes the package-level compile-registration comment above stale, because the module no longer registers all 66 concrete ops after removing AlibiFwdOp/SinusoidalFwdOp → update that comment to call out the two eager-only generative ops.

File: tileops/ops/elementwise/alibi.py [hunk 7]
```diff
@@ -40,26 +46,12 @@ def __init__(
         self.dtype = dtype
         self.dispatch_kernel(kernel_map)
```
Comment (Ibuki-wind) [inline tileops/ops/elementwise/alibi.py:47]: This direct kernel call makes AlibiFwdOp eager-only, but the class docstring still does not say that `torch.compile` graph capture is unsupported, so readers will assume the old wrapper remains available → add a short eager-only note to the docstring.

File: tileops/ops/elementwise/sinusoidal.py [hunk 8]
```diff
@@ -40,26 +46,12 @@ def __init__(
         self.dtype = dtype
         self.dispatch_kernel(kernel_map)
```
Comment (Ibuki-wind) [inline tileops/ops/elementwise/sinusoidal.py:47]: This direct kernel call makes SinusoidalFwdOp eager-only, but the class docstring still does not say that `torch.compile` graph capture is unsupported, so readers will assume the old wrapper remains available → add a short eager-only note to the docstring.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
The validator change still needs one regression test before merge.

File: scripts/validate_manifest.py (hunk 5 above)
Comment (lcy-seso) [reply scripts/validate_manifest.py:294]: Adopted. Added `test_signature_outputs_empty_fails` covering the L0 `outputs_count < 1` branch. See 15d40e2.

File: tileops/ops/elementwise/__init__.py (hunk 6 above)
Comment (lcy-seso) [reply tileops/ops/elementwise/__init__.py:279]: Adopted. Updated the registration-section comment to call out that AlibiFwdOp / SinusoidalFwdOp are intentionally excluded (zero tensor inputs, eager-only). See 40bf0bb.

File: tileops/ops/elementwise/alibi.py (hunk 7 above)
Comment (lcy-seso) [reply tileops/ops/elementwise/alibi.py:47]: Adopted. Added an eager-only / no-torch.compile-wrapping Note to the AlibiFwdOp docstring. See 40bf0bb.

File: tileops/ops/elementwise/sinusoidal.py (hunk 8 above)
Comment (lcy-seso) [reply tileops/ops/elementwise/sinusoidal.py:47]: Adopted. Added an eager-only / no-torch.compile-wrapping Note to the SinusoidalFwdOp docstring. See 40bf0bb.

**Review by Ibuki-wind — APPROVED** [review]
Clean — no issues.
